{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7e661939",
   "metadata": {},
   "source": [
    "# Notes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cc1ddb0",
   "metadata": {},
   "source": [
    "Used simplified version of the data used in the paper \"Embeddings from protein language models predict conservation and variant effects\" by Marquet et al.\n",
    "\n",
    "**Data explanation**:\n",
    "- `seq_and_conservation.txt`: FASTA file with 3 lines per entry: amino acid sequence, continuous conservation scores (ignored) and conservation classes between 1 and 9 (1 = very variable, 9 = very conserved)\n",
    "- `train_ids.txt`: IDs of the proteins for training (9392 proteins)\n",
    "- `val_ids.txt`: IDs of the proteins for validation (555 proteins)\n",
    "- `test_ids.txt`: IDs of the proteins for testing (519 proteins)\n",
    "\n",
    "**Proposed splits**:\n",
    "- `sampled`: Randomly split sequences into `train`/`test` with 95/5% probability.\n",
    "\n",
    "This is a well-known dataset used to validate the behavior of code and models. Only provided a `sampled` split for this purpose."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d900b85",
   "metadata": {},
   "source": [
    "# Configs & Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5dbd1522",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "from pandas import DataFrame, read_json\n",
    "import json\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8322f978",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Where the raw data is stored and where processed data will be deposited\n",
    "data_path = Path('') / '..' / 'data' / 'conservation'\n",
    "\n",
    "sequences = data_path / 'seq_and_conservation.txt'\n",
    "train = data_path / 'train_ids.txt'\n",
    "validation = data_path / 'val_ids.txt'\n",
    "test = data_path / 'test_ids.txt'\n",
    "\n",
    "split_path = Path('') / '..' / 'splits' / 'conservation'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cdabf48",
   "metadata": {},
   "source": [
    "# Obtain original dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0a66fc36",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = SeqIO.parse(sequences, \"fasta\")\n",
    "\n",
    "seqs_and_cons = list(parser)\n",
    "\n",
    "dataset = DataFrame(columns = [\"id\", \"sequence\", \"conservations\"])\n",
    "for i in range(0, len(seqs_and_cons), 3):\n",
    "    id = seqs_and_cons[i].description[1:].replace(' ', '-')\n",
    "    sequence = seqs_and_cons[i].seq\n",
    "    conservations = seqs_and_cons[i+2].seq\n",
    "    \n",
    "    dataset = dataset.append({\"id\": id,\n",
    "                              \"sequence\": str(sequence), \n",
    "                              \"conservations\": str(conservations)}, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ee664989",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>conservations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3p6z-C</td>\n",
       "      <td>AHHHHHHVGTWENLYFQSIPDDDEDSYEIFEPPESTVMATRKMHDR...</td>\n",
       "      <td>6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,5,7,7,3,4,7,3,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6fb3-A</td>\n",
       "      <td>TGSLVSLIRGQVVTTDGTPLVGVNVSFVKYPKYGYTITRQDGMFDL...</td>\n",
       "      <td>6,4,9,8,6,7,9,6,9,9,7,9,3,8,1,7,7,6,8,8,6,9,8,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2oqg-A</td>\n",
       "      <td>QGMTVGTYAELASVFAALSDETRWEILTELGRADQSASSLATRLPV...</td>\n",
       "      <td>1,6,6,7,5,5,4,1,6,3,4,6,2,6,8,5,9,8,7,9,5,8,9,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3ix7-A</td>\n",
       "      <td>SNAPRGGKVLDTSVLVDGRVAEVAAVGFLEGPLWVPHFVLKELQHF...</td>\n",
       "      <td>6,4,1,1,1,1,1,8,5,8,9,9,9,7,8,9,9,9,9,8,3,7,6,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5y9q-A</td>\n",
       "      <td>GTISIGCSSLIGQTLLPEVLSLYNAQFPNVEIQVQVGSTEQIKANH...</td>\n",
       "      <td>9,8,9,7,7,8,7,9,8,7,6,8,7,8,5,9,9,1,6,9,6,1,7,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10502</th>\n",
       "      <td>1ws8-A</td>\n",
       "      <td>MATVHKVGDSTGWTTLVPYDYAKWASSNKFHVGDSLLFNYNNKFHN...</td>\n",
       "      <td>8,9,6,3,6,4,9,9,6,2,1,7,9,3,2,4,3,2,3,5,7,3,3,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10503</th>\n",
       "      <td>4qa8-A</td>\n",
       "      <td>GAMDPEFGKKPTTASSPSPGSPSPEAQQILQDSSKATKGLHSVHVV...</td>\n",
       "      <td>7,6,4,4,6,4,1,5,1,1,1,1,3,3,3,3,1,1,2,1,1,1,6,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10504</th>\n",
       "      <td>1v2z-A</td>\n",
       "      <td>STAFFFRRMSPADKRKLLDELRSIYRTIVLEYFNTDAKVNERIDEF...</td>\n",
       "      <td>6,7,1,1,6,2,6,3,7,1,1,1,4,3,1,1,1,1,1,1,7,4,1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10505</th>\n",
       "      <td>3h7h-B</td>\n",
       "      <td>GSHMDPNLWTVKCKIGEERATAISLMRKFIAYQFTDTPLQIKSVVA...</td>\n",
       "      <td>6,9,1,3,9,8,6,8,8,5,7,7,8,5,2,9,6,9,6,2,6,6,2,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10506</th>\n",
       "      <td>3noq-A</td>\n",
       "      <td>GSHMAVQIGFLLFPEVQQLDLTGPHDVLASLPDVQVHLIWKEPGPV...</td>\n",
       "      <td>7,6,7,5,3,4,4,8,7,5,6,5,6,7,5,6,7,5,7,9,6,5,7,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10507 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id                                           sequence  \\\n",
       "0      3p6z-C  AHHHHHHVGTWENLYFQSIPDDDEDSYEIFEPPESTVMATRKMHDR...   \n",
       "1      6fb3-A  TGSLVSLIRGQVVTTDGTPLVGVNVSFVKYPKYGYTITRQDGMFDL...   \n",
       "2      2oqg-A  QGMTVGTYAELASVFAALSDETRWEILTELGRADQSASSLATRLPV...   \n",
       "3      3ix7-A  SNAPRGGKVLDTSVLVDGRVAEVAAVGFLEGPLWVPHFVLKELQHF...   \n",
       "4      5y9q-A  GTISIGCSSLIGQTLLPEVLSLYNAQFPNVEIQVQVGSTEQIKANH...   \n",
       "...       ...                                                ...   \n",
       "10502  1ws8-A  MATVHKVGDSTGWTTLVPYDYAKWASSNKFHVGDSLLFNYNNKFHN...   \n",
       "10503  4qa8-A  GAMDPEFGKKPTTASSPSPGSPSPEAQQILQDSSKATKGLHSVHVV...   \n",
       "10504  1v2z-A  STAFFFRRMSPADKRKLLDELRSIYRTIVLEYFNTDAKVNERIDEF...   \n",
       "10505  3h7h-B  GSHMDPNLWTVKCKIGEERATAISLMRKFIAYQFTDTPLQIKSVVA...   \n",
       "10506  3noq-A  GSHMAVQIGFLLFPEVQQLDLTGPHDVLASLPDVQVHLIWKEPGPV...   \n",
       "\n",
       "                                           conservations  \n",
       "0      6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,5,7,7,3,4,7,3,...  \n",
       "1      6,4,9,8,6,7,9,6,9,9,7,9,3,8,1,7,7,6,8,8,6,9,8,...  \n",
       "2      1,6,6,7,5,5,4,1,6,3,4,6,2,6,8,5,9,8,7,9,5,8,9,...  \n",
       "3      6,4,1,1,1,1,1,8,5,8,9,9,9,7,8,9,9,9,9,8,3,7,6,...  \n",
       "4      9,8,9,7,7,8,7,9,8,7,6,8,7,8,5,9,9,1,6,9,6,1,7,...  \n",
       "...                                                  ...  \n",
       "10502  8,9,6,3,6,4,9,9,6,2,1,7,9,3,2,4,3,2,3,5,7,3,3,...  \n",
       "10503  7,6,4,4,6,4,1,5,1,1,1,1,3,3,3,3,1,1,2,1,1,1,6,...  \n",
       "10504  6,7,1,1,6,2,6,3,7,1,1,1,4,3,1,1,1,1,1,1,7,4,1,...  \n",
       "10505  6,9,1,3,9,8,6,8,8,5,7,7,8,5,2,9,6,9,6,2,6,6,2,...  \n",
       "10506  7,6,7,5,3,4,4,8,7,5,6,5,6,7,5,6,7,5,7,9,6,5,7,...  \n",
       "\n",
       "[10507 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26de46f1",
   "metadata": {},
   "source": [
    "# Splits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e1999aef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's create a sequences.fasta file with all the sequences\n",
    "with open(split_path / 'splits' / 'sequences.fasta', 'w') as sequences_file:\n",
    "    for index, row in dataset.iterrows():\n",
    "        sequences_file.write('>{}\\n'.format(row['id']))\n",
    "        sequences_file.write('{}\\n'.format(row['sequence']))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f22ec8a",
   "metadata": {},
   "source": [
    "## sampled"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "48fc1b48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Obtain IDs of Train, Validation and Test proteins\n",
    "with open(train) as train_file:\n",
    "    train_ids = [line.rstrip() for line in train_file]\n",
    "\n",
    "with open(validation) as validation_file:\n",
    "    validation_ids = [line.rstrip() for line in validation_file]\n",
    "    \n",
    "with open(test) as test_file:\n",
    "    test_ids = [line.rstrip() for line in test_file]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "663ef72d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's create the split dataset\n",
    "sampled = DataFrame(columns = [\"id\", \"conservations\", \"set\", \"validation\"])\n",
    "\n",
    "# Add train samples\n",
    "tmp = dataset[dataset[\"id\"].isin(train_ids)][[\"id\", \"conservations\"]]\n",
    "tmp.insert(2, \"set\", [\"train\"]*len(tmp))\n",
    "sampled = sampled.append(tmp)\n",
    "\n",
    "# Add validation samples\n",
    "tmp = dataset[dataset[\"id\"].isin(validation_ids)][[\"id\", \"conservations\"]]\n",
    "tmp.insert(2, \"set\", [\"train\"]*len(tmp))\n",
    "tmp.insert(2, \"validation\", [True]*len(tmp))\n",
    "sampled = sampled.append(tmp)\n",
    "\n",
    "# Add test samples\n",
    "tmp = dataset[dataset[\"id\"].isin(test_ids)][[\"id\", \"conservations\"]]\n",
    "tmp.insert(2, \"set\", [\"test\"]*len(tmp))\n",
    "sampled = sampled.append(tmp)\n",
    "\n",
    "# Target to str\n",
    "sampled[\"conservations\"] = sampled[\"conservations\"].apply(lambda x: ''.join(map(str, x)))\n",
    "sampled[\"conservations\"] = sampled[\"conservations\"].apply(lambda x: x.replace(',', ''))\n",
    "\n",
    "sampled = sampled.reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a156151b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>conservations</th>\n",
       "      <th>set</th>\n",
       "      <th>validation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3p6z-C</td>\n",
       "      <td>6666666666666666577347367897785511671549996336...</td>\n",
       "      <td>train</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6fb3-A</td>\n",
       "      <td>6498679699793817768869898782213149394992896997...</td>\n",
       "      <td>train</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2oqg-A</td>\n",
       "      <td>1667554163462685987958983762584125368548813516...</td>\n",
       "      <td>train</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3ix7-A</td>\n",
       "      <td>6411111858999789999837631668458154782998199928...</td>\n",
       "      <td>train</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5y9q-A</td>\n",
       "      <td>9897787987687859916961751179286768397595572334...</td>\n",
       "      <td>train</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10461</th>\n",
       "      <td>1vyb-A</td>\n",
       "      <td>6552654567486978867643884563436544457746699897...</td>\n",
       "      <td>test</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10462</th>\n",
       "      <td>4co6-A</td>\n",
       "      <td>8647745748566562341194165155674439135615526992...</td>\n",
       "      <td>test</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10463</th>\n",
       "      <td>3biy-A</td>\n",
       "      <td>6317638516287116919881171251111531926988675591...</td>\n",
       "      <td>test</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10464</th>\n",
       "      <td>1sdi-A</td>\n",
       "      <td>6886125876999767987179666953932411331455196525...</td>\n",
       "      <td>test</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10465</th>\n",
       "      <td>3io3-A</td>\n",
       "      <td>5514175572563133366567789999999994878684368583...</td>\n",
       "      <td>test</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10466 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id                                      conservations    set  \\\n",
       "0      3p6z-C  6666666666666666577347367897785511671549996336...  train   \n",
       "1      6fb3-A  6498679699793817768869898782213149394992896997...  train   \n",
       "2      2oqg-A  1667554163462685987958983762584125368548813516...  train   \n",
       "3      3ix7-A  6411111858999789999837631668458154782998199928...  train   \n",
       "4      5y9q-A  9897787987687859916961751179286768397595572334...  train   \n",
       "...       ...                                                ...    ...   \n",
       "10461  1vyb-A  6552654567486978867643884563436544457746699897...   test   \n",
       "10462  4co6-A  8647745748566562341194165155674439135615526992...   test   \n",
       "10463  3biy-A  6317638516287116919881171251111531926988675591...   test   \n",
       "10464  1sdi-A  6886125876999767987179666953932411331455196525...   test   \n",
       "10465  3io3-A  5514175572563133366567789999999994878684368583...   test   \n",
       "\n",
       "      validation  \n",
       "0            NaN  \n",
       "1            NaN  \n",
       "2            NaN  \n",
       "3            NaN  \n",
       "4            NaN  \n",
       "...          ...  \n",
       "10461        NaN  \n",
       "10462        NaN  \n",
       "10463        NaN  \n",
       "10464        NaN  \n",
       "10465        NaN  \n",
       "\n",
       "[10466 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampled"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "437a258a",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(split_path / 'splits' / 'sampled.fasta', 'w') as sampled_file:\n",
    "    for index, row in sampled.iterrows():\n",
    "        validation = 'True' if row['validation'] == True else 'False'\n",
    "        sampled_file.write('>{}\\n'.format('{} SET={} VALIDATION={}'.format(row['id'], row['set'], validation)))\n",
    "        sampled_file.write('{}\\n'.format(row['conservations']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31c4d8e1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
